### PR TITLE
[Cache] Fix key encoding issue in Memcached adapter

### PR DESCRIPTION
**Pull request by lstrojny**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | n.A.
| License       | MIT
| Doc PR        | Fix double encoding in memcached which lead to overlong keys being generated

Because the memcached adapter uses `rawurlencode()` to encode each and every key, keys will sometimes be too long and therefore hit the memcached limit of 250 bytes. This happens when the key is small enough to be below 250 when passed to `AbstractAdapterTrait::getId()` and is then blown up over the 250 bytes limit in memcached adapter without validating the length again.

Looking through the code, it seems that the double encoding is wholly unnecessary assuming if one makes sure the namespace is properly encoded. This PR therefore removes the double encoding and instead uses rawurlencode on the namespace (which is in turn properly accounted for when calculating whether or not we are over the ID limit).

### CODE REVIEW

**Review by nicolas-grekas — COMMENTED** [review]
Thanks for the PR.
I understand that rawurlencode increases the length of keys, but encoding them is nonetheless required since memcached doesn't allow spaces in them, while we do.
This breaks it currently, isn't it?

File: src/Symfony/Component/Cache/Tests/Adapter/AdapterTestCase.php [hunk 1]
```diff
@@ -104,7 +104,7 @@ public function testGetMetadata()
             $this->markTestSkipped($this->skippedTests[__FUNCTION__]);
         }
 
-        $cache = $this->createCachePool(0, __FUNCTION__);
```
[the remark this answers is not in the document]
Comment (lstrojny) [reply src/Symfony/Component/Cache/Tests/Adapter/AdapterTestCase.php]: Done

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php [hunk 2]
```diff
@@ -42,11 +42,11 @@ public static function setUpBeforeClass(): void
         }
     }
 
-    public function createCachePool(int $defaultLifetime = 0): CacheItemPoolInterface
+    public function createCachePool(int $defaultLifetime = 0, string $namespace = null): CacheItemPoolInterface
```
[the remark this answers is not in the document]
Comment (lstrojny) [reply src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: Done

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php [hunk 3]
```diff
@@ -248,4 +248,26 @@ public function testMultiServerDsn()
         ];
         $this->assertSame($expected, $client->getServerList());
     }
+
+    public function testDoubleEncodingBug()
+    {
+        $pool = $this->createCachePool(0, 'contains spaces and /'.random_int(0, \PHP_INT_MAX));
+
+        // Max ID length diveded by 3 so that rawurlencode would bring us over the limit
+        $key = str_repeat('%', 250 / 3);
```
[the remark this answers is not in the document]
Comment (lstrojny) [reply src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: It doesn’t matter really as it needs to be a) small enough to be less than 250 when it passes through the SF caching subsystem but >250 after rawurlencode()

**Comment by lstrojny** [comment]
> Thanks for the PR.
> I understand that rawurlencode increases the length of keys, but encoding them is nonetheless required since memcached doesn't allow spaces in them, while we do.
> This breaks it currently, isn't it?

Great point, indeed, spaces would now break. So what I did now is adding a custom encoding function that replaces memcached reserved characters with Symfony reserved characters but keep the length of the key.

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php [hunk 4]
```diff
@@ -248,4 +248,26 @@ public function testMultiServerDsn()
         ];
         $this->assertSame($expected, $client->getServerList());
     }
+
+    public function testDoubleEncodingBug()
+    {
+        $pool = $this->createCachePool(0, 'contains spaces and /'.random_int(0, \PHP_INT_MAX));
+
+        // Max ID length diveded by 3 so that rawurlencode would bring us over the limit
```
[the remark this answers is not in the document]
Comment (lstrojny) [reply src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php:256]: Added a better explanation for the chosen key

File: src/Symfony/Component/Cache/Traits/MemcachedTrait.php [hunk 5]
```diff
@@ -31,6 +31,19 @@ trait MemcachedTrait
         \Memcached::OPT_SERIALIZER => \Memcached::SERIALIZER_PHP,
     ];
 
+    /**
+     * We are replacing characters that are illegal in Memcached keys with reserved characters from
+     * {@see \Symfony\Contracts\Cache\ItemInterface::RESERVED_CHARACTERS} that are legal in Memcached.
```
[the remark this answers is not in the document]
Comment (lstrojny) [reply src/Symfony/Component/Cache/Traits/MemcachedTrait.php:36]: Unfortunately yes 😁

File: src/Symfony/Component/Cache/Traits/MemcachedTrait.php [hunk 6]
```diff
@@ -31,6 +31,19 @@ trait MemcachedTrait
...
+     */
+    private static $encodingTable = [
```
[the remark this answers is not in the document]
Comment (lstrojny) [reply src/Symfony/Component/Cache/Traits/MemcachedTrait.php]: Done

File: src/Symfony/Component/Cache/Traits/MemcachedTrait.php [hunk 7]
```diff
@@ -31,6 +31,12 @@ trait MemcachedTrait
         \Memcached::OPT_SERIALIZER => \Memcached::SERIALIZER_PHP,
     ];
 
+    /**
+     * We are replacing characters that are illegal in Memcached keys with reserved characters from
+     * {@see \Symfony\Contracts\Cache\ItemInterface::RESERVED_CHARACTERS} that are legal in Memcached.
+     */
+    private static $encoding = ['memcached' => " \n\r\t\v\f", 'symfony' => '@():{}'];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/MemcachedTrait.php:38]: let's use two private consts :)

File: src/Symfony/Component/Cache/Traits/MemcachedTrait.php [hunk 8]
```diff
@@ -56,7 +69,7 @@ private function init(\Memcached $client, string $namespace, int $defaultLifetim
             $this->lazyClient = $client;
         }
 
-        parent::__construct($namespace, $defaultLifetime);
+        parent::__construct(rawurlencode($namespace), $defaultLifetime);
```
[the remark this answers is not in the document]
Comment (lstrojny) [reply src/Symfony/Component/Cache/Traits/MemcachedTrait.php:65]: Not in this case. Here we need a key that is compatible with both Symfony and Memcached

File: src/Symfony/Component/Cache/Traits/MemcachedTrait.php (hunk 7 above)
Comment (lstrojny) [reply src/Symfony/Component/Cache/Traits/MemcachedTrait.php:38]: Traits cannot have constants

File: src/Symfony/Component/Cache/Traits/MemcachedTrait.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Traits/MemcachedTrait.php:65]: then can't we just remove the call to rawurlencode?

File: src/Symfony/Component/Cache/Traits/MemcachedTrait.php (hunk 7 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Traits/MemcachedTrait.php:38]: let's use two private static vars, we'll turn them into consts in v5, where the trait has been removed

File: src/Symfony/Component/Cache/Traits/MemcachedTrait.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Traits/MemcachedTrait.php]: I removed the call to `rawurlencode()`: the namespace cannot contain invalid PSR-6 chars

File: src/Symfony/Component/Cache/Traits/MemcachedTrait.php (hunk 7 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Traits/MemcachedTrait.php]: updated

**Comment by nicolas-grekas** [comment]
LGTM but there are some failures to investigate Github Actions.

**Comment by lstrojny** [comment]
@nicolas-grekas tests pass now, fixed further encoding issue. I forced push over your changes but restored the variable rename. I've seen that you removed the rawurlencode() for the namespace and I expanded the testcase to demonstrate why I think it’s needed. Can you elaborate on why you think it’s not needed?

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php [hunk 9]
```diff
@@ -248,4 +249,36 @@ public function testMultiServerDsn()
         ];
         $this->assertSame($expected, $client->getServerList());
     }
+
+    public function testDoubleEncodingBug()
+    {
+        $reservedMemcachedCharacters = " \n\r\t\v\f\0";
+
+        $namespace = ItemInterface::RESERVED_CHARACTERS.$reservedMemcachedCharacters.random_int(0, \PHP_INT_MAX);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php:257]: namespaces cannot contain reserved characters: that's a restriction that is shared by all adapters, no reason to make this one different

Comment (lstrojny) [reply src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php:257]: @nicolas-grekas fixed

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @lstrojny.

**Comment by nicolas-grekas** [comment]
Arg, github action [just failed](https://github.com/symfony/symfony/runs/1090346616) with 
```
1) Symfony\Component\Cache\Tests\Simple\MemcachedCacheTextModeTest::testGetMultiple
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'foo'
+'value'

/home/runner/work/symfony/symfony/vendor/cache/integration-tests/src/SimpleCacheTest.php:331
```

**Comment by lstrojny** [comment]
@nicolas-grekas it's a randomly failing test, try this:

```
for x in {0..1000} ; do echo $x ; simple-phpunit --filter "testGetMultiple$" src/Symfony/Component/Cache/Tests/Simple/MemcachedCacheTextModeTest.php || break ; done
```

From a first glimpse it looks like `getMulti()` does not always return in order

**Comment by nicolas-grekas** [comment]
> From a first glimpse it looks like getMulti() does not always return in order

oh, then we might want to make tests resilient to this. If you want to give it a try, PR welcome :)

**Comment by lstrojny** [comment]
@nicolas-grekas correction, it is indeed the patch. It happens when the namespace version contains a slash. This seems to me to be a violation of "internal protocol" in the sense that an adapter should expect to not deal with keys containing PSR-6 reserved characters. WDYT?

**Comment by nicolas-grekas** [comment]
Adapters can deal with non-psr6 chars actually. We might have a conflict with versioning then. Would using `:` work instead of the slash?
Alternatively, we could make parsing for the version more robust, to prevent the conflict, if that makes sense (I don't remember and I didn't check :) )

**Comment by lstrojny** [comment]
This would fix it:
```
diff --git a/src/Symfony/Component/Cache/Traits/AbstractTrait.php b/src/Symfony/Component/Cache/Traits/AbstractTrait.php
index d4a16959a8..d865bd30cb 100644
--- a/src/Symfony/Component/Cache/Traits/AbstractTrait.php
+++ b/src/Symfony/Component/Cache/Traits/AbstractTrait.php
@@ -119,7 +119,7 @@ trait AbstractTrait
                 }
             }
             $namespaceToClear = $this->namespace.$namespaceVersionToClear;
-            $namespaceVersion = substr_replace(base64_encode(pack('V', mt_rand())), static::NS_SEPARATOR, 5);
+            $namespaceVersion = self::createNewNamespaceVersion();
             try {
                 $cleared = $this->doSave([static::NS_SEPARATOR.$this->namespace => $namespaceVersion], 0);
             } catch (\Exception $e) {
@@ -268,7 +268,7 @@ trait AbstractTrait
                     $this->namespaceVersion = $v;
                 }
                 if ('1'.static::NS_SEPARATOR === $this->namespaceVersion) {
-                    $this->namespaceVersion = substr_replace(base64_encode(pack('V', time())), static::NS_SEPARATOR, 5);
+                    $this->namespaceVersion = self::createNewNamespaceVersion();
                     $this->doSave([static::NS_SEPARATOR.$this->namespace => $this->namespaceVersion], 0);
                 }
             } catch (\Exception $e) {
@@ -300,4 +300,9 @@ trait AbstractTrait
     {
         throw new \DomainException('Class not found: '.$class);
     }
+
+    private static function createNewNamespaceVersion(): string
+    {
+        return substr_replace(bin2hex(pack('V', mt_rand())), static::NS_SEPARATOR, 7);
+    }
 }
```

**Comment by lstrojny** [comment]
I am not sure why previously time() and mt_rand() was used in two different places.

**Comment by nicolas-grekas** [comment]
I won't be able to have a look today, but please submit a PR and I will tomorrow.

**Comment by lstrojny** [comment]
Done, see https://github.com/symfony/symfony/pull/38126

**Pull request merged**